Address rare case of NaN test statistics for AD GoF

adewit opened the pull request:
Proposed solution for #699. There is a corresponding CH PR (https://github.com/cms-analysis/CombineHarvester/pull/272) to ensure -ve test statistic values are not picked up when collecting the results

adewit commented:
@amarini I assume your comment "I'm not convinced this is the right approach. In general, it could be problematic to exclude toys from the gof. Do we know what is the limit value of the ts when we approach zero?" was meant for this one (or the associated CH PR, I could change it from `<0` to `<-0.5` to be safe)

The test statistic should never be negative, so this only comes about when we manually set it to -1, which happens when the test statistic would have been NaN as the toy is basically faulty (no data generated). In that case I think we should treat it as a failed toy. I don't think there's an alternative - setting the test statistic value to 0 would be even worse, as people might not notice something is wrong with some of the toys.

amarini commented:
Yes, this is the one I wanted to comment :D ...
I understand that we set it up because it was nan, but what I'm meaning is that in (P=Poisson) P(data=0, exp=lambda=0) we have that the expression is nan, (e^-0 0^0 / 0!), but we now that the solution exists and it is taken has exp=0, and the negLogL could be expanded to lambda and the limit on that is obtained for lambda->0 to include the solution. In this case it is probably not correct to exclude those toys from the test-statistics because the method we have limitations (the minimization should fail since it is at the boundary and the functional is also nan ...), but we should incorporate them with the extrapolated value, as we do for the likelihood estimation and all the related stuff.  If I read correctly this case of 0 observation was exactly one of the problematic behind this PR, right?

adewit commented:
@amarini Yes, this is for the case of 0 events in the generated toy dataset. So are you saying it should be included with test stat value 0? My argument for excluding these toys is that IMO things are not well defined here, so better not to include them (the point being that if I have some weird model where 95% of the toys end up having 0 obs, then I'll just get a large peak at 0 and I might conclude that my p-value is really low if it lies somewhere >0, when actually the answer is that I can't really perform the GoF test in this way)

Happy to take other opinions too (though note the original discussion started 3 months ago on #699, so if someone has an opinion please shout soon :-) )

ajgilbert commented:
I think in this case the change to combine is ok - because this is particularly about the KS & AD tests, where the empirical distribution function is not defined if nData = 0. For the CH PR, perhaps excluding values with value < -0.5 would be safer, and maybe we can print a message like: 'X/Y toys have negative test statistic values, and are excluded. This might indicate a failure in the calculation within combine, or for the KS and AD tests, an undefined value in toys with zero events. Note that the resulting p-value could be biased.' ?

amarini commented:
For me it is fine to add a warning on screen.

adewit commented:
Ok, I've updated the comments in the docs, now print a warning when an undefined test statistic is identified, and adjusted the collection script. I think this should be everything but if anyone would like to change something else let me know. Otherwise I think we can merge.
